### PR TITLE
Improve report about costs

### PR DESCRIPTION
**Pull request by GoPavel**
# Make synthesis report more clear and self-contains.

## Problem 1: 
Currently it prints only the best solution and groups it by item (i.g. by
`(no_production, position_of_dot)`. It's incontinent when you are investigating
a specific decision of the algorithm in a specific state.

Solution: group solutions by states and print all potential solutions sorted by
cost

## Problem 2: 
Current report contains information only about solutions, but nothing
about how it comes. To reduce the requirement for source code knowledge (in some
cases handling recovery attributes is tricky) it would be better to make the
report more verbose.

Solution: print non-trivial cost of (non)terminal symbols, productions and items
penalty

### CODE REVIEW

**Comment by GoPavel** [comment]
Example of report:
```

(* * COSTS OF SOLUTIONS: *)

State: #0

Item(  0, 0) at cost 110:
 s' ::= . s
 => [Seq [Seq [Seq [Seq [Shift (T RP)]; Reduce 3]; Seq [Shift (T LP)]; Seq [Shift (T RP)]; Reduce 1]; Seq [Shift (T EOF)]; Reduce 4]; Reduce 0]


State: #1

Item(  3, 1) at cost 101:
 b ::= RP .
 => [Reduce 3]


State: #2

Item(  0, 1) at cost  1:
 s' ::= s .
 => [Reduce 0]


State: #3

Item(  1, 1) at cost  4:
 a ::= b . LP RP
 => [Seq [Shift (T LP)]; Seq [Shift (T RP)]; Reduce 1]


State: #4

Item(  1, 2) at cost  3:
 a ::= b LP . RP
 => [Seq [Shift (T RP)]; Reduce 1]


State: #5

Item(  1, 3) at cost  1:
 a ::= b LP RP .
 => [Reduce 1]


State: #6

Item(  4, 1) at cost  2:
 s ::= a . EOF
 => [Seq [Shift (T EOF)]; Reduce 4]
Item(  2, 1) at cost 101:
 b ::= a .
 => [Reduce 2]


State: #7

Item(  4, 2) at cost  1:
 s ::= a EOF .
 => [Reduce 4]



(* * COST OF PRODUCTIONS (not equal one): *)

Production b ::= RP
 at cost 101
Production b ::= a
 at cost 101

(* * COST OF TERMINAL (not equal one) : *)

error at cost inf
RP at cost  2

(* * COST OF NONTERMINAL (not equal inf) : *)


(* * PENALTY OF ITEMS (not equal zero): *)

Item (  3, 1) at cost 100
Item (  2, 1) at cost 100
```
Grammar:
```

%token LP  [@recover.cost 0]
%token RP  [@recover.cost 1]
%token EOF [@recover.cost 0]

%start <int> s

%%
 
s: a EOF {$1};

a: b  LP RP  { $1 + 1 };

b [@recover.cost 100]:
| a { $1 }
| RP  { 0  };
```

File: menhir-recover/synthesis.ml [hunk 1]
```diff
@@ -290,18 +287,62 @@ struct
...
+          fprintf ppf "%a at cost %s\n"
+              Print.nonterminal n
+              (cost_to_string @@ cost_of_symbol (N n))
+    );
+    fprintf ppf "\n(* * PENALTY OF ITEMS (not equal zero): *)\n\n";
+    Lr0.iter (fun lr0 ->
+      List.iter (fun ((prod, pos) as item) ->
+         let cost = penalty_of_item item in
+         if (Cost.compare cost Cost.zero) <> 0 then
+             fprintf ppf "Item (%3d, %d) at cost %s\n"
```
Comment (XJIE6) [inline menhir-recover/synthesis.ml:343]: Use `Print.item`. `prod` id is also useful for better navigation

File: menhir-recover/synthesis.ml [hunk 2]
```diff
@@ -290,18 +287,62 @@ struct
...
+             (sort_by_cost items);
+        fprintf ppf "\n\n"
+      ) solutions;
+    fprintf ppf "\n(* * COST OF PRODUCTIONS (not equal one): *)\n\n";
+    Lr1.iter (fun lr1 ->
+      List.iter (fun (prod, pos) ->
+        let prod_len = Array.length (Production.rhs prod) in
+        let cost = cost_of_prod prod in
+        if pos = prod_len && Cost.compare cost (Cost.of_int 1) <> 0 then
+            fprintf ppf "Production %a at cost %s\n"
```
Comment (XJIE6) [inline menhir-recover/synthesis.ml:317]: Add `prod` id for better navigation

**Review by XJIE6 — CHANGES_REQUESTED** [review]

File: menhir-recover/synthesis.ml [hunk 3]
```diff
@@ -290,18 +287,62 @@ struct
...
+                Print.production prod
+                (cost_to_string cost)
+      ) (Lr0.items (Lr1.lr0 lr1))
+    );
+    fprintf ppf "\n(* * COST OF TERMINAL (not equal one) : *)\n\n";
+    Terminal.iter (fun t ->
+      let cost = cost_of_symbol (T t) in
+      if (Cost.compare cost (Cost.of_int 1)) <> 0 then
+        fprintf ppf "%a at cost %s\n"
```
[the remark this answers is not in the document]
Comment (GoPavel) [reply menhir-recover/synthesis.ml:326]: It's an internal token of menhir. It's implicitly added by menhir always.

File: menhir-recover/synthesis.ml [hunk 4]
```diff
@@ -290,18 +287,62 @@ struct
...
+    List.iter (fun (state, items) ->
+        fprintf ppf "State: #%d\n" (Lr1.to_int state);
+        fprintf ppf "\n";
+        List.iter (fun (item, solution) ->
+                print_item_with_solution ppf (item, solution))
+             (sort_by_cost items);
+        fprintf ppf "\n\n"
+      ) solutions;
+    fprintf ppf "\n(* * COST OF PRODUCTIONS (not equal one): *)\n\n";
+    Lr1.iter (fun lr1 ->
```
[the remark this answers is not in the document]
Comment (GoPavel) [reply menhir-recover/synthesis.ml]: fixed

File: menhir-recover/synthesis.ml (hunk 2 above)
Comment (GoPavel) [reply menhir-recover/synthesis.ml]: fixed

File: menhir-recover/synthesis.ml [hunk 5]
```diff
@@ -290,18 +287,62 @@ struct
...
+    fprintf ppf "\n(* * COST OF NONTERMINAL (not equal inf) : *)\n\n";
+    Nonterminal.iter (fun n ->
+        let cost = cost_of_symbol (N n) in
+        if not (Cost.is_infinite cost) then
+          fprintf ppf "%a at cost %s\n"
+              Print.nonterminal n
+              (cost_to_string @@ cost_of_symbol (N n))
+    );
+    fprintf ppf "\n(* * PENALTY OF ITEMS (not equal zero): *)\n\n";
+    Lr0.iter (fun lr0 ->
```
[the remark this answers is not in the document]
Comment (GoPavel) [reply menhir-recover/synthesis.ml]: fixed

File: menhir-recover/synthesis.ml (hunk 1 above)
Comment (GoPavel) [reply menhir-recover/synthesis.ml]: fixed

**Comment by XJIE6** [comment]
https://gitlab.com/serokell/ligo/ligo/-/merge_requests/228

**Comment by GoPavel** [comment]
Thanks for your corrections!

**Comment by GoPavel** [comment]
Example of report for last version:
Grammar:
```

%token LP  [@recover.cost 0]
%token RP  [@recover.cost 1]
%token EOF [@recover.cost 0]

%start <int> s

%%
 
s [@recover.expr 0]: a EOF {$1};

a: b [@recover.cost 42] LP RP  { $1 + 1 };

b [@recover.cost 100]:
| a { $1 }
| RP  { 0  };
```
Report (a new part of the output that is produced by `-v` option):
```
...
# LR(1) state #0

Items:
s' ::= . s
Transitions:
 - on RP, goto #1
 - on s, goto #2
 - on b, goto #3
 - on a, goto #6
Reductions:

# LR(1) state #1

Items:
b ::= RP .
Transitions:
Reductions:
 - on LP, reduce 3:
  b ::= RP


# LR(1) state #2

Items:
s' ::= s .
Transitions:
Reductions:
 - on #, reduce 0:
  s' ::= s


# LR(1) state #3

Items:
a ::= b . LP RP
Transitions:
 - on LP, goto #4
Reductions:

# LR(1) state #4

Items:
a ::= b LP . RP
Transitions:
 - on RP, goto #5
Reductions:

# LR(1) state #5

Items:
a ::= b LP RP .
Transitions:
Reductions:
 - on LP, reduce 1:
  a ::= b LP RP

 - on EOF, reduce 1:
  a ::= b LP RP


# LR(1) state #6

Items:
s ::= a   . EOF
b ::= a .
Transitions:
 - on EOF, goto #7
Reductions:
 - on LP, reduce 2:
  b ::= a


# LR(1) state #7

Items:
s ::= a EOF .
Transitions:
Reductions:
 - on #, reduce 4:
  s ::= a EOF


# Production p0
s' ::= s

# Production p1
a ::= b LP RP

# Production p2
b ::= a

# Production p3
b ::= RP

# Production p4
s ::= a EOF



(* COSTS OF SOLUTIONS: *)

State: #0

Item(  0, 0) at cost  2:
 s' ::= . s
 => [Seq [Shift (N s)]; Reduce 0]


State: #1

Item(  3, 1) at cost 101:
 b ::= RP .
 => [Reduce 3]


State: #2

Item(  0, 1) at cost  1:
 s' ::= s .
 => [Reduce 0]


State: #3

Item(  1, 1) at cost  4:
 a ::= b . LP RP
 => [Seq [Shift (T LP)]; Seq [Shift (T RP)]; Reduce 1]


State: #4

Item(  1, 2) at cost  3:
 a ::= b LP . RP
 => [Seq [Shift (T RP)]; Reduce 1]


State: #5

Item(  1, 3) at cost  1:
 a ::= b LP RP .
 => [Reduce 1]


State: #6

Item(  4, 1) at cost  2:
 s ::= a . EOF
 => [Seq [Shift (T EOF)]; Reduce 4]
Item(  2, 1) at cost 101:
 b ::= a .
 => [Reduce 2]


State: #7

Item(  4, 2) at cost  1:
 s ::= a EOF .
 => [Reduce 4]



(* COST OF PRODUCTIONS (not equal one): *)

Production#2 b ::= a
 at cost 101
Production#3 b ::= RP
 at cost 101

(* COST OF TERMINAL (not equal one): *)

error at cost inf
RP at cost  2

(* COST OF NONTERMINAL (not equal inf): *)

s at cost  1

(* PENALTY OF ITEMS (not equal zero): *)

Item (  1, 0) a ::= . b LP RP
 at cost 42
Item (  2, 1) b ::= a .
 at cost 100
Item (  3, 1) b ::= RP .
 at cost 100
```

**Pull request merged**